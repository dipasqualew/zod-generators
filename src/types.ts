/**
 * Represents a seed value
 */
export type Seed = number | boolean | null;

export type RNG = () => number;

/**
 * Recursively applies Partial
 * to all keys in T
 */
export type DeepPartial<T> = {
  [P in keyof T]?: DeepPartial<T[P]>;
};


/**
 * Returns an object that maps
 * all the keys of T to unknown
 */
export type RequiredKeys<T> = {
  [K in keyof Required<T>]: unknown;
};

/**
 * Returns an object that partially maps
 * all the keys of T to unknown
 */
export type PartialKeys<T> = {
  [K in keyof T]?: unknown
};


/**
 * Defines the shape of the options
 * that can be passed to a mock function
 */
export interface MockOptions<T> {
  override?: DeepPartial<T> | (() => DeepPartial<T>);
  seed?: Seed;
  rng?: RNG;
}

/**
 * Defines a collection of mocks
 */
export type Mocks<T> = Record<string, T>;

/**
 * Defines a mocks
 */
export type MocksGenerator = <T>() => Mocks<T>;
