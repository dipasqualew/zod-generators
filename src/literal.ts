import type { ZodLiteral } from 'zod';

/**
 * Generates valid literal mock
 * from the given ZodLiteral definition
 *
 * @param field
 */
export const mockValid = <T>(field: ZodLiteral<T>) => {
  return {
    DEFAULT: field._def.value,
  };
};

/**
 * Generates invalid literal mocks
 * from the given ZodLiteral definition
 *
 * @param field
 */
export const mockInvalid = <T>(field: ZodLiteral<T>) => {
  return {
    // This will catch all values
    // and transform it to something else
    DEFAULT: !field._def.value,
  };
};
