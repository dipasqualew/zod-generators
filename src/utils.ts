/**
 * Latin alphabet
 */
export const ALPHABET = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789'.split('');

export const MIN_INTEGER = Math.round(Number.MIN_SAFE_INTEGER / 3);
export const MAX_INTEGER = Math.round(Number.MAX_SAFE_INTEGER / 3);

/**
 * Returns a random element
 * from the given array
 *
 * @param origin
 */
export const getRandomElement = <T>(origin: T[]): T => {
  const index = Math.floor(Math.random() * origin.length);
  const item = origin[index];

  return item;
};

/**
 * Returns a random number between
 * the given min and the given max
 *
 * @param min
 * @param max
 * @param integer Whether the number should be an integer
 */
export const getRandomNumber = (min: number, max: number, integer = false): number => {
  const safeMin = Math.max(min, MIN_INTEGER);
  const safeMax = Math.min(max, MAX_INTEGER);

  const num = (Math.random() * (safeMax - safeMin)) + safeMin;

  if (integer) {
    return Math.floor(num);
  }

  return num;
};


/**
 * Returns a random string
 * of the given length
 *
 * @param length
 * @param alphabet
 */
export const getString = (length: number, alphabet = ALPHABET): string => {
  const elements = Array.from({ length }).map(() => getRandomElement(alphabet));

  return elements.join('');
};

/**
 * Returns a random string with
 * the given minLength and maxLength
 *
 * @param minLength
 * @param maxLength
 * @param alphabet
 */
export const getRandomString = (minLength: number, maxLength: number, alphabet = ALPHABET): string => {
  const length = getRandomNumber(minLength, maxLength, true);

  return getString(length, alphabet);
};
