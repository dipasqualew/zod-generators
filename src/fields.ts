/* eslint-disable no-use-before-define */
import * as zod from 'zod';

import * as literal from './literal';
import * as any from './primitives/any';
import * as bigint from './primitives/bigint';
import * as boolean from './primitives/boolean';
import * as date from './primitives/date';
import * as znull from './primitives/null';
import * as number from './primitives/number';
import * as string from './primitives/string';
import * as zundefined from './primitives/undefined';

/**
 * Mock generator function
 */
export type MockGenerator = () => { DEFAULT?: unknown } & Record<string, unknown>;

/**
 * Returns the generator of valid values
 * for the given field
 *
 * @param field
 */
export const getValidGenerator = <T extends zod.ZodAny>(field: T): MockGenerator => {
  if (field instanceof zod.ZodString) {
    return () => string.mockValid(field);
  }
  if (field instanceof zod.ZodNumber) {
    return () => number.mockValid(field);
  }

  if (field instanceof zod.ZodBoolean) {
    return () => boolean.mockValid(field);
  }

  if (field instanceof zod.ZodArray) {
    return () => generateValidArrays(field);
  }

  if (field instanceof zod.ZodObject) {
    return () => generateValidObjects(field as zod.ZodObject<zod.ZodRawShape>);
  }

  if (field instanceof zod.ZodDate) {
    return () => date.mockValid(field);
  }

  if (field instanceof zod.ZodBigInt) {
    return () => bigint.mockValid(field);
  }

  if (field instanceof zod.ZodLiteral) {
    return () => literal.mockValid(field);
  }

  if (field instanceof zod.ZodNull) {
    return () => znull.mockValid(field);
  }

  if (field instanceof zod.ZodUndefined || field instanceof zod.ZodVoid) {
    return () => zundefined.mockValid(field);
  }

  return () => any.mockValid(field);
};

/**
 * Returns the generator of invalid values
 * for the given field
 *
 * @param field
 */
export const getInvalidGenerator = <T extends zod.ZodAny>(field: T): MockGenerator => {
  if (field instanceof zod.ZodString) {
    return () => string.mockInvalid(field);
  }

  if (field instanceof zod.ZodNumber) {
    return () => number.mockInvalid(field);
  }

  if (field instanceof zod.ZodBoolean) {
    return () => boolean.mockInvalid(field);
  }

  if (field instanceof zod.ZodArray) {
    return () => generateInvalidArrays(field);
  }

  if (field instanceof zod.ZodObject) {
    return () => generateInvalidObjects(field as zod.ZodObject<zod.ZodRawShape>);
  }

  if (field instanceof zod.ZodDate) {
    return () => date.mockInvalid(field);
  }

  if (field instanceof zod.ZodLiteral) {
    return () => literal.mockInvalid(field);
  }

  if (field instanceof zod.ZodBigInt) {
    return () => bigint.mockInvalid(field);
  }

  if (field instanceof zod.ZodNull) {
    return () => znull.mockInvalid(field);
  }

  if (field instanceof zod.ZodUndefined || field instanceof zod.ZodVoid) {
    return () => zundefined.mockInvalid(field);
  }

  return () => any.mockInvalid(field);
};

/**
 * Returns the valid mocks
 * for the given field
 *
 * @param field
 */
export const mockValid = <T extends zod.ZodAny>(field: T) => {
  const generator = getValidGenerator(field);

  return generator();
};

/**
 * Returns the invalid mocks
 * for the given field
 *
 * @param field
 */
export const mockInvalid = <T extends zod.ZodAny>(field: T) => {
  const generator = getInvalidGenerator(field);

  return generator();
};

/**
 * Returns all mocks
 * for the given field
 *
 * @param field
 */
export const mock = <T extends zod.ZodAny>(field: T) => {
  return {
    valid: mockValid(field),
    invalid: mockInvalid(field),
  };
};

/**
 * Generates valid arrays
 * from the given ZodArray definition
 *
 * @param field
 */
export const generateValidArrays = <T extends zod.ZodAny>(field: zod.ZodArray<T>) => {
  const generator = getValidGenerator(field._def.type);
  const items = generator();

  return {
    DEFAULT: Object.values(items),
  };
};

/**
 * Generates invalid arrays
 * from the given ZodArray definition
 *
 * @param field
 */
export const generateInvalidArrays = <T extends zod.ZodAny>(field: zod.ZodArray<T>) => {
  const generator = getInvalidGenerator(field._def.type);
  const items = generator();

  const values = Object.values(items);

  if (values.length) {
    return {
      DEFAULT: values,
    };
  }

  return {};
};

/**
 * Generates valid objects
 * from the given ZodObject definition
 *
 * @param field
 */
export const generateValidObjects = <T extends zod.ZodRawShape>(field: zod.ZodObject<T>) => {
  type Shape = zod.infer<typeof field>;

  const defaultMock: Partial<Shape> = {};
  const variations: Record<string, Record<string, unknown>> = {};

  Object.entries(field.shape).forEach((entry) => {
    const key = entry[0] as keyof Shape;
    const def = entry[1] as Shape[keyof Shape];
    const generator = getValidGenerator(def);
    const validMocks = generator();

    defaultMock[key] = validMocks.DEFAULT as Shape[keyof Shape];
    variations[entry[0]] = validMocks;
  });

  const mocks = [
    ['DEFAULT', defaultMock],
  ];

  Object.entries(variations).forEach((variation) => {
    Object.entries(variation[1]).forEach((entry) => {
      const name = `${variation[0]}__${entry[0]}`;
      const value = { ...defaultMock, [variation[0]]: entry[1] };

      mocks.push([name, value]);
    });
  });

  return Object.fromEntries(mocks) as { DEFAULT: Shape } & Record<string, Shape>;
};

/**
 * Generates valid objects
 * from the given ZodObject definition
 *
 * @param field
 */
export const generateInvalidObjects = <T extends zod.ZodRawShape>(field: zod.ZodObject<T>) => {
  type Shape = zod.infer<typeof field>;

  const valid = generateValidObjects(field).DEFAULT;

  const defaultMock: Partial<Shape> = {};
  const variations: Record<string, Record<string, unknown>> = {};

  Object.entries(field.shape).forEach((entry) => {
    const key = entry[0] as keyof Shape;
    const def = entry[1] as Shape[keyof Shape];
    const generator = getInvalidGenerator(def);
    const invalidMocks = generator();

    if ('DEFAULT' in invalidMocks) {
      defaultMock[key] = invalidMocks.DEFAULT as Shape[keyof Shape];
      variations[entry[0]] = invalidMocks;
    }
  });

  const mocks: [string, unknown][] = [];

  if (Object.keys(defaultMock).length) {
    mocks.push(['DEFAULT', defaultMock]);
  }

  Object.entries(variations).forEach(([fieldName, fieldVariations]) => {
    Object.entries(fieldVariations).forEach(([description, fieldValue]) => {
      const name = `${fieldName}__${description}`;
      const value = { ...valid, [fieldName]: fieldValue };

      mocks.push([name, value]);
    });
  });

  return Object.fromEntries(mocks) as { DEFAULT: unknown } & Record<string, unknown>;
};
