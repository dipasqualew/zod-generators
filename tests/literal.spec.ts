import { mockInvalid, mockValid } from '../src/literal';
import { literalFields } from './fields';
import { invalidateGroup, validateGroup } from './mixins';

describe('literal', () => {
  literalFields.forEach(([description, field]) => {
    describe(description, () => {
      describe('validates', () => {
        const valid = mockValid(field);
        validateGroup(field, valid);
      });

      describe('invalidates', () => {
        const valid = mockInvalid(field);
        invalidateGroup(field, valid);
      });
    });
  });
});
