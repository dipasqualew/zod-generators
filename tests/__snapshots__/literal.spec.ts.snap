// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`literal a boolean literal invalidates Invalidates DEFAULT 1`] = `
"[
  {
    \\"code\\": \\"invalid_type\\",
    \\"expected\\": true,
    \\"received\\": false,
    \\"path\\": [],
    \\"message\\": \\"Expected true, received false\\"
  }
]"
`;

exports[`literal a number literal invalidates Invalidates DEFAULT 1`] = `
"[
  {
    \\"code\\": \\"invalid_type\\",
    \\"expected\\": 1265,
    \\"received\\": false,
    \\"path\\": [],
    \\"message\\": \\"Expected 1265, received false\\"
  }
]"
`;

exports[`literal a string literal invalidates Invalidates DEFAULT 1`] = `
"[
  {
    \\"code\\": \\"invalid_type\\",
    \\"expected\\": \\"Dante Alighieri\\",
    \\"received\\": false,
    \\"path\\": [],
    \\"message\\": \\"Expected Dante Alighieri, received false\\"
  }
]"
`;
